src/eth_rpc/servers/kakarot_rpc.rs

```diff
@@ -0,0 +1,97 @@
+    pub const fn new(eth_provider: EP, starknet_provider: SP) -> Self {
+        Self { eth_provider, starknet_provider }
+    }
+}
+trait ToElements {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, eyre::Error>;
+}
+
+impl ToElements for BroadcastedInvokeTransaction {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, eyre::Error> {
+        match self {
+            Self::V1(tx_v1) => Ok(tx_v1),
+            Self::V3(_) => Err(eyre::eyre!("Transaction is V3, cannot convert to V1")),
+        }
+    }
+}
+
+#[async_trait]
+impl<EP, SP> KakarotApiServer for KakarotRpc<EP, SP>
+where
+    EP: EthereumProvider + Send + Sync + 'static,
+    SP: Provider + Send + Sync + 'static,
+{
+    async fn get_starknet_transaction_hash(&self, hash: B256, retries: u8) -> RpcResult<B256> {
+        // Retrieve the stored transaction from the database.
+        let transaction = self
+            .eth_provider
+            .transaction_by_hash(hash)
+            .await
+            .map_err(ErrorObject::from)?
+            .ok_or_else(|| ErrorObject::owned(INVALID_PARAMS_CODE, "Transaction not found", None::<()>))?;
+
+        // Convert the `Transaction` instance to a `TransactionSigned` instance.
+        let transaction_signed_ec_recovered: reth_primitives::TransactionSignedEcRecovered = transaction
+            .try_into()
+            .map_err(|_| ErrorObject::owned(INVALID_PARAMS_CODE, "Failed to convert transaction", None::<()>))?;
+
+        let (transaction_signed, _) = transaction_signed_ec_recovered.to_components();
+
+        // Retrieve the signer of the transaction.
+        let signer = transaction_signed
+            .recover_signer()
+            .ok_or_else(|| ErrorObject::owned(INVALID_PARAMS_CODE, "Failed to recover signer", None::<()>))?;
+        // Create the Starknet transaction.
+        let starknet_transaction = to_starknet_transaction(&transaction_signed, signer, retries)
+            .map_err(|_| {
+                ErrorObject::owned(INVALID_PARAMS_CODE, "Failed to convert to StarkNet transaction", None::<()>)
+            })?
+            .try_into_v1()
+            .map_err(|_| {
+                ErrorObject::owned(
+                    INVALID_PARAMS_CODE,
+                    "Failed to convert StarkNet transaction to version 1",
+                    None::<()>,
+                )
+            })?;
+
+        let chain_id = self.starknet_provider.chain_id().await.unwrap();
+
+        // Compute the hash on elements
+        let transaction_hash = compute_hash_on_elements(&[
+            Felt::from_bytes_be_slice(b"invoke"),
+            Felt::ONE,
+            starknet_transaction.sender_address,
+            Felt::ZERO,
+            compute_hash_on_elements(&starknet_transaction.calldata),
+            starknet_transaction.max_fee,
+            chain_id,
+            starknet_transaction.nonce,
+        ]);
+
+        Ok(B256::from_slice(&transaction_hash.to_bytes_be()[..]))
+    }
+}
```

Comment:
I don't know if you agree but I would do something more like (untested):

```Rust
trait ToElements {
    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, EthApiError>;
}

impl ToElements for BroadcastedInvokeTransaction {
    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, EthApiError> {
        match self {
            Self::V1(tx_v1) => Ok(tx_v1),
            Self::V3(_) => Err(EthApiError::EthereumDataFormat(EthereumDataFormatError::TransactionConversion)),
        }
    }
}

#[async_trait]
impl<EP, SP> KakarotApiServer for KakarotRpc<EP, SP>
where
    EP: EthereumProvider + Send + Sync + 'static,
    SP: Provider + Send + Sync + 'static,
{
    async fn get_starknet_transaction_hash(&self, hash: B256, retries: u8) -> RpcResult<Option<B256>> {
        // Retrieve the stored transaction from the database.
        let transaction = self.eth_provider.transaction_by_hash(hash).await?;

        if let Some(transaction) = transaction {
            // Convert the `Transaction` instance to a `TransactionSigned` instance.
            let transaction_signed_ec_recovered: reth_primitives::TransactionSignedEcRecovered = transaction
                .try_into()
                .map_err(|_| EthApiError::EthereumDataFormat(EthereumDataFormatError::TransactionConversion))?;

            let (transaction_signed, _) = transaction_signed_ec_recovered.to_components();

            // Retrieve the signer of the transaction.
            let signer = transaction_signed.recover_signer().ok_or(SignatureError::Recovery)?;

            // Create the Starknet transaction.
            let starknet_transaction = to_starknet_transaction(&transaction_signed, signer, retries)?.try_into_v1()?;

            let chain_id = self.starknet_provider.chain_id().await.unwrap();

            // Compute the hash on elements
            let transaction_hash = compute_hash_on_elements(&[
                Felt::from_bytes_be_slice(b"invoke"),
                Felt::ONE,
                starknet_transaction.sender_address,
                Felt::ZERO,
                compute_hash_on_elements(&starknet_transaction.calldata),
                starknet_transaction.max_fee,
                chain_id,
                starknet_transaction.nonce,
            ]);

            Ok(Some(B256::from_slice(&transaction_hash.to_bytes_be()[..])))
        }

        Ok(None)
    }
}
```

Some comments on this proposal:
- I avoid putting error codes directly in but rather go through conversions to `RpcResult`
- As `transaction_by_hash` returns an `Option`, that is to say potentially a `None`, for me we must also reflect this behavior here and if the `transaction_by_hash` returns a `None` then we also return a `None` here.

Let me know what you think about this @greged93 @eugypalu

Reply:
makes sense to me, I'll add it. thanks